lib/spack/spack/test/spec_semantics.py

```diff
@@ -785,3 +789,49 @@ def test_spec_flags_maintain_order(self):
                 s.compiler_flags[x] == ['-O0', '-g']
                 for x in ('cflags', 'cxxflags', 'fflags')
             )
+
+    def test_any_combination_of(self):
+        # Test that using 'none' and another value raise during concretization
+        spec = Spec('multivalue_variant foo=none,bar')
+        with pytest.raises(spack.error.SpecError) as exc_info:
+            spec.concretize()
+
+        assert "is mutually exclusive with any of the" in str(exc_info.value)
+
+    @pytest.mark.skipif(
+        sys.version_info[0] == 2, reason='__wrapped__ requires python 3'
+    )
+    def test_errors_in_variant_directive(self):
+        variant = spack.directives.variant.__wrapped__
+
+        class Pkg(object):
+            name = 'PKG'
+
+        # We can't use names that are reserved by Spack
+        fn = variant('patches')
+        with pytest.raises(spack.directives.DirectiveError) as exc_info:
+            fn(Pkg())
+        assert "The name 'patches' is reserved" in str(exc_info.value)
+
+        # We can't have conflicting definitions for arguments
+        fn = variant(
+            'foo', values=spack.variant.any_combination_of('fee', 'foom'),
+            default='bar'
```

Comment:
@alalazo when I first skimmed this I missed that this is not failing because `'bar'` doesn't exist but rather because `variant.any_combination_of` is made responsible for specifying the default.

We've talked before about the error message and I like it, but when I examine this particular test I get the feeling there is a conflict of sorts between the `DisjointSets` constructor logic (using the fluent interface) and the typical `variant()` declaration logic.

I think it would be totally fine to throw this error if the user did something like:

```
fn = variant(
  'foo',
  values=spack.variant.any_combination_of('fee', 'foom')
    .with_default('fee'),
  default='fee'
)
```

But if the user didn't explicitly specify `with_default`, then ideally setting the `default=` argument within the `variant` would override the implicit default chosen by the `DisjointSets` constructor.

I think this is more of an issue for `'default'` rather than for `'multi'` or `'validator'`

Does that complaint make sense? I think that minimally it is worth adding a TODO for it.

Reply:
TODO added